File: Core/Algorithms/LongestSubstring/longest-substring.py
"""
Time Complexity: O(n * m)
Space Complexity: O(m)
Where 'n' is the length of the string and
'm' is the total number of unique characters in the string
"""

# def lengthOfLongestSubstringBrute(s: str):
#     res = 0
#     for i in range(len(s)):
#         charSet = set()
#         for j in range(i, len(s)):
#             if s[j] in charSet:
#                 break
#             charSet.add(s[j])
#         res = max(res, len(charSet))
#     return res


"""
    Sliding Window
    TC: O(n)
    SC: O(m)
    'n' is the length of the string and
    'm' is the total number of unique characters in the string
"""


def lengthOfLongestSubstring(s: str) -> int:
    charSet = set()  # tracks unique chars in current window
    left = 0  # left pointer
    maxLength = 0  # best result so far

    for right in range(len(s)):  # r is the right pointer, always moves forward
        while s[right] in charSet:  # duplicate found — window is invalid
            charSet.remove(s[left])  # remove the leftmost char
            left += 1  # shrink window from the left
        charSet.add(s[right])  # safe to add — no duplicate now
        maxLength = max(maxLength, right - left + 1)  # window size = r - l + 1
    return maxLength


# print(lengthOfLongestSubstring("abcabcbb"))
# print(lengthOfLongestSubstring("zxyzxyz"))
# print(lengthOfLongestSubstring("xxxx"))
# print(lengthOfLongestSubstring("pwwkew"))

File: Core/Algorithms/LongestSubstring/LongestSubstring.java
package GoldmanSachs.LongestSubstring;

import java.util.HashSet;

/**
 * Brute force: for each char position, check every substring if it has any duplicate,
 * if it doesn't, return the length of the longest substring
 * O(n^2)
 */
public class LongestSubstring {
    /**
     * Sliding Window
     * Finds the length of the longest substring without repeating characters.
     * Time Complexity: O(N) - Each character is visited at most twice (once by 'right', once by 'left').
     * Space Complexity: O(min(N, M)) - N is string length, M is the size of the character set (alphabet).
     */
    public int lengthOfLongestSubstring(String s) {
        // 1. Validation: Handle null or empty strings early
        if (s == null || s.isEmpty()) {
            return 0;
        }

        // 2. Data Structure Setup:
        // We use a HashSet as a "sliding window" tracker to check for duplicates in O(1) time.
        HashSet<Character> visitedChars = new HashSet<>();
        int left = 0;    // The 'start' of our current unique substring
        int maxLength = 0;

        /*
        * 3. The Sliding Window Strategy:
        * The 'right' pointer expands the window by exploring new characters.
        */
        for (int right = 0; right < s.length(); right++) {
            char currentChar = s.charAt(right);

            /*
            * 4. Violation Handling:
            * If the character at 'right' is already in our set, the window is no longer unique.
            * We shrink the window from the 'left' until the duplicate is removed.
            */
            while (visitedChars.contains(currentChar)) {
                // Remove the character at the left pointer and move left inward
                visitedChars.remove(s.charAt(left));
                left++;
            }

            // 5. Update:
            // Add the new unique character and recalculate the maximum length found so far.
            visitedChars.add(currentChar);

            // (right - left + 1) gives us the current window size
            maxLength = Math.max(maxLength, right - left + 1);
        }

        return maxLength;
    }
}
